pkg/gadgets/trace/dns/tracer/tracer.go

```diff
@@ -60,10 +60,12 @@ func NewTracer() (*Tracer, error) {
 			return nil, err
 		}
 
-		// Derive latency from the request/response timestamps.
-		if event.Qr == types.DNSPktTypeQuery {
+		// Derive latency from the query/response timestamps.
+		// Filter by packet type (OUTGOING for queries and HOST for responses) to exclude cases where
+		// the packet is forwarded between containers in the host netns.
+		if bpfEvent.Qr == 0 && bpfEvent.PktType == pktTypeOutgoing {
 			latencyCalc.storeDNSQueryTimestamp(netns, bpfEvent.Id, bpfEvent.Timestamp)
-		} else {
+		} else if bpfEvent.Qr == 1 && bpfEvent.PktType == pktTypeHost {
```

Comment:
I wonder if we can get the two following latencies by just filtering for OUTGOING/HOST (excluding ) regardless of the `Qr` field:
1. in the target pod, time between the request is sent and the reply is received
2. in the coredns pod, time between the request is received and the reply is sent (that latency should be smaller than the previous one)

I have not tested if the output would make sense in that case.

I think we might need the interface (`eth0`, `docker0`) to distinguish duplicated entries on the host network namespace. It should be available from `(struct __sk_buff *skb)->ifindex`. Then we have [net.InterfaceByIndex(ifindex)](https://pkg.go.dev/net#InterfaceByIndex)->Name but it should be called from the right netns (with `NetnsEnter()`) and it should be cached, so that should be for another PR.

Reply:
> I wonder if we can get the two following latencies by just filtering for OUTGOING/HOST (excluding ) regardless of the `Qr` field:
> 
>     1. in the target pod, time between the request is sent and the reply is received
> 
>     2. in the coredns pod, time between the request is received and the reply is sent (that latency should be smaller than the previous one)
> 
> I have not tested if the output would make sense in that case.

The behavior with the `Qr` check is:

1. if coredns queries an upstream DNS server, latency for the query from coredns->upstream is reported.
2. if coredns serves the query directly, then latency is not reported for the coredns pod (since it doesn't send any query).

I think if we removed the `Qr` check, then the behavior is the same for case (1), but for case (2) we'd instead report the latency for coredns to serve the request.

My feeling is that it's clearer if latency always means the time between a query/reply, and I'd imagine that's what someone investigating a DNS issue would care most about.
 
> I think we might need the interface (`eth0`, `docker0`) to distinguish duplicated entries on the host network namespace. It should be available from `(struct __sk_buff *skb)->ifindex`. Then we have [net.InterfaceByIndex(ifindex)](https://pkg.go.dev/net#InterfaceByIndex)->Name but it should be called from the right netns (with `NetnsEnter()`) and it should be cached, so that should be for another PR.

That makes sense. I wonder if we could skip `InterfaceByIndex` and just make the DNS cache key (netns, ifindex, dnsID).